### PR TITLE
Remove uses of calculateMinimumCompatVersion from handshakes

### PR DESCRIPTION
**Pull request by thecoop**
Don't use `computeMinimumCompatVersion` in transport handshakes. This removes one use of TransportVersion.computeMinimumCompatVersion, changing it to hard-coded constants for now. There will be a later piece of work to separate out the handshake infrastructure from the rest of the messaging code.

This is a simpler version of #93312 that doesn't introduce new types

Related #93225

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java [hunk 1]
```diff
@@ -203,23 +205,23 @@ private void ensureOpen() {
...
-        // once the connection is established
-        final TransportVersion compatibilityVersion = isHandshake ? currentVersion.calculateMinimumCompatVersion() : currentVersion;
-        if (remoteVersion.isCompatible(compatibilityVersion) == false) {
-            final TransportVersion minCompatibilityVersion = isHandshake
-                ? compatibilityVersion
-                : compatibilityVersion.calculateMinimumCompatVersion();
-            String msg = "Received " + (isHandshake ? "handshake " : "") + "message from unsupported version: [";
-            return new IllegalStateException(msg + remoteVersion + "] minimal compatible version is: [" + minCompatibilityVersion + "]");
+    static void checkHandshakeVersionCompatibility(TransportVersion handshakeVersion) {
+        if (handshakeVersion.before(TransportHandshaker.EARLIEST_HANDSHAKE_VERSION)) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:209]: I would prefer to check here that the handshake version is one of 6.8.0 or 7.17.0.

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java [hunk 2]
```diff
@@ -203,23 +205,23 @@ private void ensureOpen() {
...
+                "Received message from unsupported version: ["
+                    + handshakeVersion
+                    + "] minimal compatible version is: ["
+                    + TransportHandshaker.EARLIEST_HANDSHAKE_VERSION
+                    + "]"
+            );
+        }
+    }
+
+    static void checkVersionCompatibility(TransportVersion remoteVersion, TransportVersion currentVersion) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:220]: Requiring `currentVersion` here is a little weird IMO. I see why it's here: we lie about our current version in a few tests, but I think these are not tests where this check matters; moreover if we do lie about our current version then we may end up claiming to support a transport version which makes no sense to us. Not a topic for this PR I think, but something worth cleaning up independently.

**Review by DaveCTurner — COMMENTED** [review]
Nice, much simpler. I left a couple of comments.

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:209]: Updated

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:209]: looks unchanged - did you forget to push?

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java]: Yup :facepalm:

File: server/src/test/java/org/elasticsearch/transport/TransportHandshakerTests.java [hunk 3]
```diff
@@ -67,7 +67,7 @@ public void testHandshakeRequestAndResponse() throws IOException {
         long reqId = randomLongBetween(1, 10);
         handshaker.sendHandshake(reqId, node, channel, new TimeValue(30, TimeUnit.SECONDS), versionFuture);
 
-        verify(requestSender).sendRequest(node, channel, reqId, TransportVersion.CURRENT.minimumCompatibilityVersion());
+        verify(requestSender).sendRequest(node, channel, reqId, TransportHandshaker.EARLIEST_HANDSHAKE_VERSION);
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/transport/TransportHandshakerTests.java:70]: This change (and the other similar ones in this suite) moves the handshake version used from 7.17.0 to 6.8.0 which I think is not what we want. Indeed we should be able to assert somewhere that we only ever send 7.17.0-format handshake requests (in prod at least, we could reasonably have some kind of test utility to emulate the way that 6.8.0 handshake requests are created)

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java]: Is 7.18.0 going to appear at any point now?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java]: There's no plan to do so. It would be very disruptive.

**Comment by thecoop** [comment]
For some reason this change causes an increase in 4 bytes in the transport stats recorded. And it's not due to just closing the channel on exception, rather than sending an error response. I'm not sure where the extra data is coming from

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 4]
```diff
@@ -158,7 +158,7 @@ public void setUp() throws Exception {
 
         final Settings connectionSettings = connectionSettingsBuilder.build();
 
-        serviceA = buildService("TS_A", version0, clusterSettingsA, connectionSettings); // this one supports dynamic tracer updates
+        serviceA = buildService("TS_A", CURRENT_VERSION, clusterSettingsA, connectionSettings); // this one supports dynamic tracer updates
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java:161]: So, this change is a bit tricky. In these tests, the service was created with `CURRENT_VERSION.minimumCompatibleVersion()`. This was then called `minimumCompatibleVersion` *again* within `TransportHandshaker.sendHandshake`. This was causing the transport to initiate with the previous handshake version (6.8.0), with the older protocol.

In reality, it should be using CURRENT here, which is then turned into the `minimumCompatiblityVersion` *once*, resulting in version 7.17 (in this PR this is also replaced with the hardcoded `REQUEST_HANDSHAKE_VERSION`). This uses the new protocol, which has 4 extra bytes in the request & response.

**Comment by thecoop** [comment]
Found the source of the extra bytes - see comment

File: server/src/main/java/org/elasticsearch/transport/InboundDecoder.java (hunk 2 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/transport/InboundDecoder.java:219]: I've taken this out as part of https://github.com/elastic/elasticsearch/pull/93453

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 5]
```diff
@@ -217,19 +217,8 @@ private <T extends TransportRequest> void handleRequest(TcpChannel channel, Head
...
-                        () -> format(
-                            "could not send error response to handshake received on [%s] using wire format version [%s], closing channel",
-                            channel,
-                            header.getVersion()
-                        ),
-                        e
-                    );
-                    channel.close();
-                }
+                logger.warn(() -> format("error processing handshake received on [%s], closing channel", channel), e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:220]: nit: maybe mention the received handshake version in this message

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit)

**Pull request merged**